implementations/rust/ockam/ockam_vault/src/hash_impl.rs

```diff
@@ -1,32 +1,54 @@
-use crate::error::Error;
-use crate::software_vault_impl::SoftwareVaultImpl;
+use crate::software_vault::SoftwareVault;
+use crate::VaultError;
 use arrayref::array_ref;
-use ockam_vault_core::hash_vault::HashVault;
-use ockam_vault_core::secret::Secret;
-use ockam_vault_core::secret_vault::SecretVault;
-use ockam_vault_core::types::{
-    SecretAttributes, SecretType, AES128_SECRET_LENGTH, AES256_SECRET_LENGTH,
+use ockam_vault_core::{
+    HashVault, Secret, SecretAttributes, SecretType, SecretVault, AES128_SECRET_LENGTH,
+    AES256_SECRET_LENGTH,
 };
 use sha2::{Digest, Sha256};
 
-impl SoftwareVaultImpl {
-    fn hkdf_sha256_internal(
+impl HashVault for SoftwareVault {
+    fn sha256(&self, data: &[u8]) -> ockam_core::Result<[u8; 32]> {
+        let digest = Sha256::digest(data);
+        Ok(*array_ref![digest, 0, 32])
+    }
+
+    /// Compute sha256.
+    /// Salt and Ikm should be of Buffer type.
+    /// Output secrets should be only of type Buffer or AES
+    fn hkdf_sha256(
         &mut self,
         salt: &Secret,
         info: &[u8],
-        ikm: &[u8],
+        ikm: Option<&Secret>,
         output_attributes: Vec<SecretAttributes>,
-    ) -> Result<Vec<Secret>, ockam_core::Error> {
+    ) -> ockam_core::Result<Vec<Secret>> {
+        let ikm = match ikm {
+            Some(ikm) => {
+                let ikm = self.get_entry(ikm)?;
+                if ikm.key_attributes().stype == SecretType::Buffer {
+                    Ok(ikm.key().as_ref())
+                } else {
+                    Err(VaultError::InvalidKeyType.into())
+                }
+            }
+            None => Ok(&[] as &[u8]),
+        }?;
+
         let salt = self.get_entry(salt)?;
 
+        if salt.key_attributes().stype != SecretType::Buffer {
+            return Err(VaultError::InvalidKeyType.into());
+        }
+
         // FIXME: Doesn't work for secrets with size more than 32 bytes
         let okm_len = output_attributes.len() * 32;
 
         let okm = {
             let mut okm = vec![0u8; okm_len];
             let prk = hkdf::Hkdf::<Sha256>::new(Some(salt.key().as_ref()), ikm);
             prk.expand(info, okm.as_mut_slice())
-                .or(Err(Error::HkdfExpandError.into()))?;
+                .or(Err(VaultError::HkdfExpandError.into()))?;
```

Comment:
`clippy` is mad at this line, can you run it and fix up the warnings? The other warnings are similar in other areas

Reply:
@jared-s this becomes ugly https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=472d31cb68cf7ca636d43350f4dbfbc8